evm/rlp/blocks.py

```diff
@@ -1,9 +1,21 @@
 import rlp
 
+from evm.db.backends.memory import MemoryDB
+from evm.db.chain import BaseChainDB
+
 
 class BaseBlock(rlp.Serializable):
     db = None
 
+    def __init__(self, header, transactions=None, uncles=None):
+        self.db = BaseChainDB(MemoryDB()).db  # for generating transaction_root and receipt_root
```

Comment:
I'm not sure I understand the need for this.  Where is it used in the code?

Reply:
In non-stateless mode, the transaction trie is rebuilt via `BaseChainDB.add_transaction( block_header, index_key, transaction)`:
```python
    def add_transaction(self, block_header, index_key, transaction):
        transaction_db = Trie(self.db, root_hash=block_header.transaction_root)
        transaction_db[index_key] = rlp.encode(transaction)
        return transaction_db.root_hash
```
And it is triggered in `VM.add_transaction(transaction, computation)`:
https://github.com/hwwhww/py-evm/blob/ec9ff9c63ff96e484150bc62ab862aed5505a03d/evm/vm/base.py#L79

For stateless mode, `BaseBlock.db` is only used for storing transaction trie and receipt trie in `VMState` level 
during `create_block` because I'd want to remove `chaindb` from `VMState` eventually.

In `VMState.add_transaction`:
https://github.com/hwwhww/py-evm/blob/ec9ff9c63ff96e484150bc62ab862aed5505a03d/evm/vm_state.py#L223
```python
        # Get trie roots and changed key-values.
        tx_root_hash, tx_db = cls.add_trie_node_to_db(
            block.header.transaction_root,
            index_key,
            transaction,
            block.db,
        )
        receipt_root_hash, receipt_db = cls.add_trie_node_to_db(
            block.header.receipt_root,
            index_key,
            receipt,
            block.db,
        )
```

And then insert the transaction to `block.db`:
https://github.com/hwwhww/py-evm/blob/ec9ff9c63ff96e484150bc62ab862aed5505a03d/evm/vm_state.py#L247-L254
```python
    @staticmethod
    def add_trie_node_to_db(root_hash, index_key, node, db):
        """
        Add transaction or receipt to the given db.
        """
        trie_db = Trie(db, root_hash=root_hash)
        trie_db[index_key] = rlp.encode(node)
        return trie_db.root_hash, trie_db.db
```

Reply:
I'm wondering if we can take a different approach to decouple some of this.

The [`mk_receipt_root`](https://github.com/ethereum/pyethereum/blob/953a8a6cdc0bbcd8a10733ca6935525674b9c544/ethereum/common.py#L183) function in the pyethereum codebase is probably the utility we need to bring over.

Then, rather than keeping track of the trie for receipts and transactions we could very easily just rebuild it on demand each time we add a new transaction/receipt.  Thoughts?

Reply:
I did rewrite `mk_receipt_root` for debugging during implementing, it would be much cleaner if we use `mk_receipt_root` to get root. The trade-offs are:

1. Since we want to get the "current block" after each time we call `VMState.apply_transaction`, if `import_block` is called, the trie-rebuilding will insert the all the previous receipts in this block during each `VMState.apply_transaction` calling. Would it be more inefficient?

2. I find that if we keep the current code, **maybe** we don't need to store `VMState.receipts` in `VMState` level. I'm still checking if it's true, right now it seems that we only require maintaining `VMState.receipts` for rebuilding receipt trie.

If 2 is true, it won't be a big deal to maintain `VMState.receipts`.

Reply:
I'm comfortable with the overhead with rebuilding the trie each time.  If the overhead turns out to be significant we can address it at that time.

Reply:
by the way, you do what you think is best.  This idea isn't coming from me having a fully formed solution, but rather trying to further simplify the `Block` object and remove some coupling and database access.  It may very well be that it doesn't provide a worthwhile improvement.

Reply:
Totally agree your point. Either is fine with me for now, let me figure out it after more refactorings of `create_block`.